win32/service.py

```diff
@@ -0,0 +1,288 @@
+        log.info("Agent processes stopped.")
+
+        # Let's log the uptime
+        if self.start_ts is not None:
+            secs = int(time.time()-self.start_ts)
+            mins = int(secs/60)
+            hours = int(secs/3600)
+            log.info("Uptime: {0} hours {1} minutes {2} seconds".format(hours, mins % 60, secs % 60))
+
+        self.ReportServiceStatus(win32service.SERVICE_STOP_PENDING)
+
+    def SvcDoRun(self):
+        log.info('Starting service...')
+        servicemanager.LogMsg(
+            servicemanager.EVENTLOG_INFORMATION_TYPE,
+            servicemanager.PYS_SERVICE_STARTED,
+            (self._svc_name_, '')
+        )
+        self.start_ts = time.time()
+
+        # Start all services.
+        for proc in self.procs.values():
+            proc.start()
+
+        # Loop to keep the service running since all DD services are
+        # running in separate processes
+        self.running = True
+        while self.running:
+            # Restart any processes that might have died.
+            for name, proc in self.procs.iteritems():
+                if not proc.is_alive() and proc.is_enabled():
+                    log.warning("%s has died. Restarting..." % name)
+                    proc.restart()
+
+            time.sleep(SERVICE_SLEEP_INTERVAL)
+
+        servicemanager.LogMsg(
+            servicemanager.EVENTLOG_INFORMATION_TYPE,
+            servicemanager.PYS_SERVICE_STOPPED,
+            (self._svc_name_, '')
+        )
+        log.info("Service stopped.")
+
+
+class DDProcess(object):
+    """
+    Starts and monitors a Datadog process.
+    Restarts when it exits until the limit set is reached.
+    """
+    DEFAULT_MAX_RESTARTS = 5
+    _RESTART_TIMEFRAME = 3600
+
+    def __init__(self, name, command, env, enabled=True, max_restarts=None):
+        self._name = name
+        self._command = command
+        self._env = env.copy()
+        self._enabled = enabled
+        self._proc = None
+        self._restarts = deque([])
+        self._max_restarts = max_restarts or self.DEFAULT_MAX_RESTARTS
+
+    def start(self):
+        if self.is_enabled():
+            log.info("Starting %s.", self._name)
+            self._proc = psutil.Popen(
+                self._command,
+                stdout=AgentSvc.devnull,
+                stderr=AgentSvc.devnull,
+                env=self._env
+            )
+        else:
+            log.info("%s is not enabled, not starting it.", self._name)
+
+    def stop(self):
+        if self._proc is not None and self._proc.is_running():
+            log.info("Stopping %s...", self._name)
+            self._proc.terminate()
+
+            psutil.wait_procs([self._proc], timeout=3)
+
+            if self._proc.is_running():
+                log.debug("%s didn't exit. Killing it.", self._name)
+                self._proc.kill()
+
+            log.info("%s is stopped.", self._name)
+        else:
+            log.info('%s was not running.', self._name)
```

Comment:
Switched it to debug. I think it's still useful because it tells us the service didn't find the process. (more obvious than an absence of log)